### PR TITLE
[dynamo] share a subgraph tracer across fwd and bwd in autograd.Function

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111104

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -1053,12 +1056,30 @@ def call_function(
...
     ) -> "VariableTracker":
-        from . import ConstantVariable, UserFunctionVariable
+        from . import (
+            AutogradFunctionContextVariable,
+            ConstantVariable,
+            UserFunctionVariable,
+        )
         from .builder import wrap_fx_proxy
 
+        if self.value.__name__ == "trampoline_autograd_apply":
```
Comment (ezyang) [inline torch/_dynamo/variables/higher_order_ops.py:1072]: What's going on here?

**Comment by ezyang** [comment]
I can't easily tell from the PR desc: why do we even need a tracer for the backward in autograd.Function?

**Comment by zou3519** [comment]
> I can't easily tell from the PR desc: why do we even need a tracer for the backward in autograd.Function?

Voz and I discussed a bit offline. The new approach is to attempt to speculate_subgraph on a function that is both the forward and the backward.

The existing strategy for autograd.Function is for Dynamo to prove that the forward and backward are both traceable (by AOTAutograd), and then for dynamo to add a call to autograd.Function in the graph. Currently, to do so, we run speculate_subgraph on the forward(), then we run speculate_subgraph for the backward(). The problem that we're running into in the issue is that each speculate_subgraph call uses a different fx.Tracer object and that it is possible to pass Proxys from forward() to backward() via the ctx object and the backward fx.Tracer gets confused when it sees a Proxy that it didn't create.

**Comment by voznesenskym** [comment]
My new plan here is to make a VT to span across both fwd and backward, instead of separate ones as we have today. Then the tracer can live there. The spirit of this PR will stay the same, but I will do the VT shuffling in a PR before this one.

**Pull request closed without merging**